File: c++/数论/欧拉函数/线性筛.cpp
int p[MAX],phi[MAX],tot,prime[MAX];
void init_phi(int n)
{
	int i,j;
	tot=0;
	memset(prime,0,sizeof prime);
	prime[1]=phi[1]=1;
	for(i=2;i<=n;i++)
	{
		if(!prime[i])
		{
			prime[i]=p[tot++]=i;
			phi[i]=i-1;
		}
		for(j=0;i*p[j]<=n;j++)
		{
			prime[i*p[j]]=p[j];
			if(i%p[j]==0)
			{
				phi[i*p[j]]=phi[i]*p[j];
				break;
			}
			else phi[i*p[j]]=phi[i]*(p[j]-1);
		}
	}
}

File: python3/线性筛.py
p=init_array_1d(MAX,0)
prime=init_array_1d(MAX,0)
tot=0
def init(n):
    global tot
    n=int(n)
    tot=0
    prime[1]=1
    for i in range(2,n,1):
        if prime[i]==0:
            prime[i]=i
            p[tot]=i
            tot+=1
        for j in range(0,tot,1):
            if i*p[j]>n: break
            prime[i*p[j]]=p[j]
            if i%p[j]==0: break
